### ISSUE TITLE
Combat actions not sorted

### ISSUE THREAD

**Issue by bitcraft** [Tuxemon/Tuxemon#128]
This issue is about the order that events are executed during the combat phase.  Currently, they are executed by sorting all the actions during the round according to the speed of the monster that is executing them.  Below I've listed how I understand it should be....please feel free to comment and correct it.
- Running from combat will happen before any moves
- Items will happen next
- Healing will occur before attacks
- Everything else is sorted by speed of the monster

**Comment by tamashihoshi** [comment]
I opened a thread in the forum to ask others:
https://forum.tuxemon.org/thread.php?pid=979#p979

IMO:
running -> speed of monster -> items -> healing

but maybe something variable is also possible, and the above order is the usual order..

**Comment by bitcraft** [comment]
https://github.com/Tuxemon/Tuxemon/wiki/Combat-State-Overview

**Comment by ShadowApex** [comment]
> So, I'm wondering about "Use items". Should trainers have a speed value? So that if both (or more) trainers decide to use an item, the faster trainer is first?

I think in the case of two trainers both using items, we can randomly select the order. In general I don't foresee use of any items that would give a player an advantage if they use it first. Items can only be used on your own monsters (with the exception of capture devices). This certainly could change in the future though, so we should take that into account to make it easier on ourselves if we decide to change it.

**Comment by ShadowApex** [comment]
I think I've changed my mind on this actually. Trainer speed sounds good to add. That, or we take the monster's speed into account for using items.

**Comment by tamashihoshi** [comment]
> That, or we take the monster's speed into account for using items.
Would it be weird if we say: if a tuxemon is slow, you can use items faster on it?
(slow tuxemon doesnt move that much, it might be easier for the trainer to give the potion etc to the tuxemon)

**Comment by bitcraft** [comment]
Trainer speed seems like the best solution to me.

**Comment by xirsoi** [comment]
I disagree. Up to this point trainers have been the simplest of actors, adding attributes to them doesn't really make sense. Especially since adding ways to affect those attributes doesn't make sense.

I'd argue that item use should have a fixed speed (or maybe even a per-item speed attr, but that's more complicated than necessary) and be treated just like an attack action. Very fast tuxemon may be able to get an ability off before the item can be used. This keeps things simple, and simple is good where complexity adds nothing to the game play.

**Comment by bitcraft** [comment]
There will be situations in a battle where some sort of arbitration between trainers when using items is needed.  The simple approach of just letting the player always go first works for a single player game, but Tuxemon is being built for internet play, where multiple players might battle, trade, and capture each others monsters.

Its not possible now, but there will be PvP battles over the internet.  Suppose both trainers use their capture devices and they have equal likelihood of capturing the opponent's monster.  In this case, it would be unfair to make it random who goes first, and no amount of mental gymnastics to argue that it should depend on monster speed will work here to justify it. In this hypothetical situation, both players would lose their monsters (trading effectively), but it will affect their win/loss record.  It is simple to understand that the faster trainer will get to use the capture device first. 

If a player knows that trainer speed is a factor then the player will understand.  In 99% of the game, it will not matter, but in some PvP situations, it will be easier to understand trainer speed, compared to item order relying on monster speed.

Right now, matches are won or lost instantly, so the order that trainers operate at is crucial.  Perhaps it would be useful to consider matches that determine the winner/loser at the end of the round, instead of during it.  Maybe offering a "Tie" result.

As for single player mode, trainer speed still makes sense and offers some degree of strategy.  What you might consider complex another person might not.

**Comment by ShadowApex** [comment]
I think that introducing player attributes such as speed can add an interesting aspect to the game. Not only would you have the opportunity to level up and train your individual monsters, but you would also have the opportunity to level up your trainer in different ways. It also provides the benefit of deciding on which trainer gets the opportunity to use an item first, which adds another small layer of strategy to the game.

**Comment by xirsoi** [comment]
This only delays the issue of deciding who goes first in the case of a tie. A tie will become less likely, but is still possible. What do you do in the case that two players have the same speed and use an item at the same time? And don't tell me "we'll incorporate the tuxemon's speed as well", because that is still a delay.

Player attributes feels like feature creep. It complicates the system, and not just a little. Since adding a player speed will only solve this issue if there is a way to modify the player speed, there needs to be ways to increase that speed. So instead of implementing important game features, we delay dealing with what is basically a race condition by adding a new element that only hypothetically improves the game.

> Its not possible now, but there will be PvP battles over the internet. Suppose both trainers use their capture devices and they have equal likelihood of capturing the opponent's monster. In this case, it would be unfair to make it random who goes first, and no amount of mental gymnastics to argue that it should depend on monster speed will work here to justify it. In this hypothetical situation, both players would lose their monsters (trading effectively), but it will affect their win/loss record. It is simple to understand that the faster trainer will get to use the capture device first.

I am extremely dubious as to the wisdom of allowing players to steal each others Tuxemon, it reeks of cheese. It would be ridiculously unfun to play someone who just uses resilient Tuxemon that can hold out while the player throws capture devices at my party. In the end, this is a game, it needs to be fun.

> Right now, matches are won or lost instantly, so the order that trainers operate at is crucial. Perhaps it would be useful to consider matches that determine the winner/loser at the end of the round, instead of during it. Maybe offering a "Tie" result.

This is exactly the way Pokemon games work though, and changing it would be more confusing for the player than deciding that items work at a fixed speed, with some other factor as the tie-breaker. It also removes the ability for Tuxemon whose "schtick" is speed to gain any kind of advantage. Ending a tight match in your favor by choosing a move that is fast stops mattering.

If there was an entire new subsystem of player attributes that effect the game, I'd think having some kind of Speed attribute that governed item use would be fine. I've been designing a D&D inspired monster training game that has the trainer being much more involved which would use things like this.

As it stands, I think the best thing to do at this stage is incorporate something more straightforward and consider designing a player attribute sub-system later. Otherwise you risk having this weird appendage to the game design that doesn't fit.

**Comment by Lvl4Sword** [comment]
Why player speed? What other stats do the players have?
If you're playing the game in story mode, the player should go first.
If you're playing against someone else, it should be random. 
Capturing someone else's monsters? Uh.. what? That's a surefire way to ensure NO ONE wants to play the game.
Something as simple as a random coin toss ( shown on screen? ) where each player chooses their side would suffice. This makes it feel like the players have a choice, and doesn't seem horrible unfair.
Comments/questions/rebuttals?

**Comment by bitcraft** [comment]
Any features discussed here (such as capturing another players monsters) are purely theoretical....There are no plans for it, but as a community project with goals that are always shifting, it is wise to at least account for ways the game/code may be used.

**Comment by tomelssjo** [comment]
I am for fixed usage speed. Simply if it is your turn you can use the item first. The tuxemon speed kind of decides who turn is first through. I do not think there is an easy way to make it differently.

**Comment by bitcraft** [comment]
I've implemented the basics for combat actions sorting, the next steps will be to decide, in detail the exact order.  I'd like to close this issue now (or move it off the milestone list) so that there is progress for the next milestone.  I think I've covered the basics by implementing combat actions sorting by class (ie. healing, swapping, attack) and by monster speed.  No trainer speed is implemented ATM, but it is not off the books.

TLDR; merged code for sorting, lets close issues, or move it off milestone for now.
